targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8C63XX/TARGET_MCU_PSOC6_M4/ipc_rpc.cpp

```diff
@@ -0,0 +1,125 @@
+/*
+ * mbed Microcontroller Library
+ * Copyright (c) 2017-2018 Future Electronics
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ipc_rpc.h"
+#include "Mutex.h"
+#include "Semaphore.h"
+#include "mbed_assert.h"
+#include "cy_ipc_config.h"
+#include "ipc/cy_ipc_pipe.h"
+#include <stdarg.h>
+#include "platform/SingletonPtr.h"
+
+using namespace rtos;
+
+
+static SingletonPtr<Mutex>      msg_mutex;
+static SingletonPtr<Semaphore>  msg_semaphore;
+
+#define RPC_GEN     RPC_GEN_INTERFACE_IDS
+
+#include "rpc_defs.h"
+#include "rpc_api.h"
+
+#undef RPC_GEN
+
+extern "C" {
```

Comment:
why are these functions defined with C linkage here?

Reply:
I  have done it for consistency, because functions that are RPC-ed to M0 (from psoc6_utils.h) are low level ones, called from HAL implementation, so they are necessarily C API, I also wanted this mechanism to be most generic, so it meant lower common denominator, i.e. C. The third reason is that wrapper generation is a code shared between M0 and M4 cores, and for pre-compiled binaries for M0 core we are providing we have to use PSoC Creator IDE from Cypress and it is C-only.
However, you are right the the wrapper functions themselves are compiled with CPP compiler on Mbed, so only wrappers APIs have to be declared as C-linkage.